tensorly/contrib/mps_decomposition_cross.py

```diff
@@ -0,0 +1,430 @@
+
+    return (next_col_idx, fibers_list, Q_skeleton)
+
+
+def maxvol(A):
+    """ Find the rxr submatrix of maximal volume in A(nxr), n>=r
+
+            We want to decompose matrix A as
+                    A = A[:,J] * (A[I,J])^-1 * A[I,:]
+            This algorithm helps us find this submatrix A[I,J] from A, which has the largest determinant.
+            We greedily find vector of max norm, and subtract its projection from the rest of rows.
+
+    Parameters
+    ----------
+
+    A: matrix
+            The matrix to find maximal volume
+
+    Returns
+    -------
+    row_idx: list of int
+            is the list or rows of A forming the matrix with maximal volume,
+    A_inv: matrix
+            is the inverse of the matrix with maximal volume.
+
+    References
+    ----------
+    S. A. Goreinov, I. V. Oseledets, D. V. Savostyanov, E. E. Tyrtyshnikov, N. L. Zamarashkin.
+    How to find a good submatrix.Goreinov, S. A., et al.
+    Matrix Methods: Theory, Algorithms and Applications: Dedicated to the Memory of Gene Golub. 2010. 247-256.
+
+    Ali Çivril, Malik Magdon-Ismail
+    On selecting a maximum volume sub-matrix of a matrix and related problems
+    Theoretical Computer Science. Volume 410, Issues 47–49, 6 November 2009, Pages 4801-4811
+    """
+
+    (n, r) = tl.shape(A)
+
+    # The index of row of the submatrix
+    row_idx = tl.zeros(r)
+
+    # Rest of rows / unselected rows
+    rest_of_rows = tl.tensor(list(range(n)))
+    rest_of_rows = tl.int(rest_of_rows)
+
+    # Find r rows iteratively
+    i = 0
+    A_new = A
+    while i < r:
+        mask = list(range(A_new.shape[0]))
+        # Compute the square of norm of each row
+        rows_norms = tl.sum(A_new ** 2, axis=1)
+
+        # If there is only one row of A left, let's just return it. MxNet is not robust about the case.
+        if rows_norms.shape == ():
+            row_idx[i] = rest_of_rows
+            break
+
+        # If a row is 0, we delete it.
+        if any(rows_norms == 0):
+            zero_idx = tl.argmin(rows_norms,axis=0)
+            mask.pop(zero_idx)
+            rest_of_rows = rest_of_rows[mask]
+            A_new = A_new[mask,:]
+            continue
+
+        # Find the row of max norm
+        max_row_idx = tl.argmax(rows_norms, axis=0)
+        max_row = A[rest_of_rows[max_row_idx], :]
+
+        # Compute the projection of max_row to other rows
+        max_row = tl.transpose(max_row)
+        # projection a to b is computed as: <a,b> / sqrt(|a|*|b|)
+        projection = tl.dot(A_new, max_row)
+        normalization = tl.sqrt(rows_norms[max_row_idx] * rows_norms)
+        # make sure normalization vector is of the same shape of projection (causing bugs for MxNet)
+        normalization = tl.reshape(normalization, projection.shape)
+        projection = projection/normalization
+
+        # Subtract the projection from A_new:  b <- b - a * projection
+        A_new = A_new - A_new * projection.reshape(A_new.shape[0], 1)
```

Comment:
Would be better to use tl.shape(A_new) but that's mainly for tensorflow so..

Reply:
Sure. I have made sure all shape() is called as tl.shape().